sympy/matrices/expressions/tests/test_derivatives.py

```diff
@@ -0,0 +1,80 @@
+"""
+Some examples have been taken from:
+
+http://www.math.uwaterloo.ca/~hwolkowi//matrixcookbook.pdf
+"""
+from sympy import MatrixSymbol, Inverse, symbols, Determinant, Trace
+
+k = symbols("k")
+
+X = MatrixSymbol("X", k, k)
+x = MatrixSymbol("x", k, 1)
+
+A = MatrixSymbol("A", k, k)
+B = MatrixSymbol("B", k, k)
+C = MatrixSymbol("C", k, k)
+D = MatrixSymbol("D", k, k)
+
+a = MatrixSymbol("a", k, 1)
+b = MatrixSymbol("b", k, 1)
+c = MatrixSymbol("c", k, 1)
+d = MatrixSymbol("d", k, 1)
+
+
+def test_matrix_derivative_trivial_cases():
+    # Cookbook example 33:
+    assert X._eval_derivative(A) == 0
+
+
+def test_matrix_derivative_vectors_and_scalars():
+
+    # Cookbook example 69:
+    expr = x.T*a
+    assert expr._eval_derivative(x) == a
+    expr = a.T*x
+    assert expr._eval_derivative(x) == a
+
+    # Cookbook example 70:
+    expr = a.T*X*b
+    assert expr._eval_derivative(X) == a*b.T
+
+    # Cookbook example 71:
+    expr = a.T*X.T*b
+    assert expr._eval_derivative(X) == b*a.T
+
+    # Cookbook example 77:
+    expr = b.T*X.T*X*c
+    assert expr._eval_derivative(X) == X*b*c.T + X*c*b.T
+
+    # Cookbook example 78:
+    expr = (B*x + b).T*C*(D*x + d)
+    assert expr._eval_derivative(x) == B.T*C*(D*x + d) + D.T*C.T*(B*x + b)
+
+    # Cookbook example 81:
+    expr = x.T*B*x
+    assert expr._eval_derivative(x) == B*x + B.T*x
+
+    # Cookbook example 82:
+    expr = b.T*X.T*D*X*c
+    assert expr._eval_derivative(X) == D.T*X*b*c.T + D*X*c*b.T
+
+    # Cookbook example 83:
+    expr = (X*b + c).T*D*(X*b + c)
+    # TODO: parsing error
+    # assert expr._eval_derivative(X) == D*X*b*b.T + D*c*b.T + D.T*X*b*b.T + D.T*c*b.T
```

Comment:
@asmeurer this is the only failed example from the cookbook so far, parsing back from the index-summation is currently failing.

Reply:
What happens when it fails? Do you get an exception or an invalid expression or what?

Reply:
This case raises an exception, an _IndexError_ in one of the parsing arrays. This is definitely a bug.